### DISCUSSION TITLE
How to get a quick viewer update?

### DISCUSSION THREAD

**Discussion by lypwig** [bernhard-42/vscode-ocp-cad-viewer#118] [Q&A] [answered]
To update the viewer, I run the Python file via the "play" button.

But loading the build123d/cadquery dependencies can take a few seconds, which is a bit inconvenient when developing.

So I have two related questions:

Is it possible to load them once and then only execute the code that actually build the model to show?

Also is there a file watcher so the viewer is updated each time the file is saved?

**Comment by jdegenstein** [comment]
The VSCode Jupyter extension allows for use of ipython cells. When I develop build123d I put imports/setup into a cell at the top and then the model I am developing + `show` calls below. This greatly speeds up execution time, and furthermore enables an (IMHO) very important console (the "interactive window") that operates in the same scope as the viewer so you can e.g. inspect variables.

I believe the default shortcut to run the cell in VSCode is CTRL + Enter (at least on non-macos keyboards).

**Comment by lypwig** [comment]
Thank you for your answer!

But I don't use Jupyter and I don't plan to use it. Is it mandatory to have quick updates?

**Reply by jdegenstein** [reply]
To be clear the Jupyter extension is not the same as a jupyter-notebook (or whatever that is called). My only interaction with it is that it adds cells to the standard editor window. See video demo here of this (recommended) setup https://www.youtube.com/watch?v=UhUmMInlJic

**Comment by bernhard-42** [comment, marked as answer]
You need a running python instance (kernel) in the background to keep the state alive.
Jupyter does this, you can call repeatedly other cells' code and it will leverage the already imported libs.
If you don't want to use Jupyter, you can use the jupyter-mode in VS Code or the ipython extension (hoangkimlai.ipython) which uses a running ipython terminal where it pastes code you want to run.

So the answer is, you need either a running kernel (jupyter or jupyter mode) or a running interpreter (ipython extension) to get quick updates.

**Comment by lypwig** [comment]
Thank you!

I downloaded iPython and it works well, I can update the view by hitting F5.

~~Ideally it should run each time I save the file, but I have no idea how to do this. I found [RunOnSave](https://marketplace.visualstudio.com/items?itemName=emeraldwalk.RunOnSave) but I would like to avoid installing another extension.~~ edit: nevermind, pressing F5 saves the file before running it.
